### PR TITLE
Allow TreehouseAppContent code to be preloaded

### LINKED ISSUE #861: Content.preload()
For LazyColumn it’ll be a performance win to run asynchronous guest code eagerly, before the target `TreehouseView` is attached. One possible way to implement this is a `preload()` function on `Content`:

```
public interface Content {
  /**
   * Prepare the content for an upcoming call to [bind]. This may yield
   * lower latency for lazy layouts.
   * 
   * To avoid resource leaks every call to [preload] must be followed by
   * a call to [unbind], whether or not there is a call to [bind].
   */
  public fun preload(view: TreehouseView)

  ...
}
```

The `preload` function doesn’t _really_ need the `TreehouseView` yet, but accepting it keeps the `hostConfiguration` quite simple. Otherwise we’d need to accept a `HostConfiguration` and fuse it to the `StateFlow<HostConfiguration>` that follows.

### PR DESCRIPTION
**Pull request by veyndan**
Closes #861.

### CODE REVIEW

File: redwood-treehouse-host/src/commonMain/kotlin/app/cash/redwood/treehouse/TreehouseAppContent.kt [hunk 1]
```diff
@@ -163,31 +306,52 @@ private class ViewContentCodeBinding<A : AppService>(
...
+    if (diffCount++ == 0) {
+      view.reset()
+      codeListener.onCodeLoaded(view, isInitialLaunch)
 
-      bridge.sendDiff(diff)
+      // Update the TreehouseAppContent's ViewState on the first diff.
+      val currentState = stateFlow.value
+      if (currentState.codeState is CodeState.Running && currentState.codeState.viewContentCodeBinding == this) {
+        stateFlow.value = State(currentState.viewState, CodeState.Running(this, hasDiffs = true))
+      }
```
Comment (veyndan) [inline redwood-treehouse-host/src/commonMain/kotlin/app/cash/redwood/treehouse/TreehouseAppContent.kt]: @swankjesse In the spirit of just having one way of doing things, is there any material difference between `awaitContent` and `codeListener#onCodeLoaded` (apart from one being a suspend function and the other being a callback)?

I know I said it'd be nicer to have a suspend function for the `ViewLazyColumn` implementation, but making a callback suspending isn't that difficult.

Comment (veyndan) [reply redwood-treehouse-host/src/commonMain/kotlin/app/cash/redwood/treehouse/TreehouseAppContent.kt]: Actually, I think this code is in the wrong place. It's only executed if we have a `view` that's non-null, which isn't what we want. Gonna move it up.

File: redwood-treehouse-host/src/commonMain/kotlin/app/cash/redwood/treehouse/SequentialStateFlow.kt [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * limitations under the License.
+ */
+package app.cash.redwood.treehouse
+
+import kotlinx.coroutines.flow.FlowCollector
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.collectLatest
+
+public class SequentialStateFlow<T>(
```
Comment (JakeWharton) [inline redwood-treehouse-host/src/commonMain/kotlin/app/cash/redwood/treehouse/SequentialStateFlow.kt]: internal

File: redwood-treehouse-host/src/commonMain/kotlin/app/cash/redwood/treehouse/TreehouseAppContent.kt (hunk 1 above)
Comment (swankjesse) [reply redwood-treehouse-host/src/commonMain/kotlin/app/cash/redwood/treehouse/TreehouseAppContent.kt]: ooooh good catch.

**Review by swankjesse — APPROVED** [review]

File: redwood-treehouse-host/src/commonMain/kotlin/app/cash/redwood/treehouse/TreehouseAppContent.kt (hunk 1 above)
Comment (veyndan) [reply redwood-treehouse-host/src/commonMain/kotlin/app/cash/redwood/treehouse/TreehouseAppContent.kt]: https://github.com/cashapp/redwood/pull/870/commits/3ee06eba42d766bcd4354bb593ff041014599f10

**Pull request merged**